infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -98,24 +180,101 @@ std::vector<std::byte> read_memory(pid_t pid, unsigned long long address, size_t
-  if (memory.size() == 0) {
-    return;
+// Construct a string with the memory specified in a register.
+std::string read_register_string(pid_t pid, unsigned long reg, unsigned long length) {
+  auto memory = read_memory(pid, reg, length);
+  if (!memory.size()) {
+    return "";
   }
 
-  std::string path(reinterpret_cast<char*>(
-        memory.data()), std::min(memory.size(), kTripWire.length()));
+  std::string content(reinterpret_cast<char*>(
+        memory.data()), std::min(memory.size(), length));
+  return content;
+}
+
+void report_bug(int bug_code) {
+  // Report the bug found based on the bug code.
+  fprintf(stderr, "===BUG DETECTED: %s===\n", kBugMessageMap[bug_code].c_str());
+  // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
+  // to the root process.
+  // Note: this may not be reliable or consistent if shell injection happens
+  // in an async way.
+  kill(g_root_pid, SIGABRT);
+  _exit(0);
+}
+
+void inspect_for_injection(pid_t pid, const user_regs_struct &regs) {
+  // Inspect a PID's registers for the sign of shell injection.
+  std::string path = read_register_string(pid, regs.rdi, kTripWire.length());
+  if (!path.length()) {
+    return;
+  }
   debug_log("inspecting");
   if (path == kTripWire) {
-    fprintf(stderr, "===BUG DETECTED: Shell injection===\n");
-    // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
-    // to the root process.
-    // Note: this may not be reliable or consistent if shell injection happens
-    // in an async way.
-    kill(g_root_pid, SIGABRT);
-    _exit(0);
+    report_bug(kShellInjection);
+  }
+}
+
+std::string get_pathname(pid_t pid, const user_regs_struct &regs) {
+  // Parse the pathname from the memory specified in the ROI register.
+  std::string raw_content = read_register_string(pid, regs.rdi, kShellPathnameLength);
+
+  std::string pathname = raw_content.substr(0, raw_content.find(" "));
+  debug_log("Pathname is %s (len %lu)\n", pathname.c_str(), pathname.length());
+  return pathname;
+  }
+
+std::string match_shell(std::string shell_pathname) {
+  // Identify the name of the shell used in the pathname.
+  for (std::string known_shell : kShellSet) {
+    if (shell_pathname.length() < known_shell.length()) {
+      continue;
+    }
+    std::string shell = shell_pathname.substr(shell_pathname.find_last_of("/")+1, known_shell.length());
+    if (!shell.compare(known_shell)) {
+      debug_log("Matched %s\n", shell.c_str());
+      return shell;
+    }
   }
+  return "";
+}
+
+std::string get_shell(pid_t pid, const user_regs_struct &regs) {
+  // Get shell name used in a PID.
+  std::string shell_pathname = get_pathname(pid, regs);
+  if (shell_pathname.length()) {
+    return match_shell(shell_pathname);
+  } else {
+    return "";
+  }
+}
+
+void match_error_pattern(std::string buffer, std::string shell) {
+  // Identify the error pattern in the write buffer and report a bug if matched.
+  std::istringstream ss{buffer};
```

Comment:
Do we need to read line by line? The buffer is max 100 chars, so it's likely this is causing more overhead than any potential optimizations.

Can we just call `buffer.find()` here on the possible error strings?